File: leetcode/101-200/188-best-time-to-buy-and-sell-stock-iv/bench/stock_iv.c
#include <stdio.h>
#include <stdlib.h>

static long max_i64(long a, long b) { return a > b ? a : b; }

static long max_profit(long k, const long *prices, long n) {
    if (n == 0 || k == 0) {
        return 0;
    }
    if (k >= n / 2) {
        long profit = 0;
        for (long i = 1; i < n; i++) {
            if (prices[i] > prices[i - 1]) {
                profit += prices[i] - prices[i - 1];
            }
        }
        return profit;
    }

    long neg = -1000000000;
    long *buy = malloc((size_t)(k + 1) * sizeof(long));
    long *sell = malloc((size_t)(k + 1) * sizeof(long));
    for (long j = 0; j <= k; j++) {
        buy[j] = neg;
        sell[j] = 0;
    }
    for (long d = 0; d < n; d++) {
        long price = prices[d];
        for (long t = 1; t <= k; t++) {
            buy[t] = max_i64(buy[t], sell[t - 1] - price);
            sell[t] = max_i64(sell[t], buy[t] + price);
        }
    }
    long ans = sell[k];
    free(buy);
    free(sell);
    return ans;
}

int main(void) {
    long n = 2000;
    long kmax = 50;
    long rounds = 5000;

    long *prices = malloc((size_t)n * sizeof(long));
    long state = 12345;
    for (long i = 0; i < n; i++) {
        state = (state * 1103515245 + 12345) & 2147483647;
        prices[i] = (state >> 16) % 1000;
    }

    long sink = 0;
    for (long round = 0; round < rounds; round++) {
        state = (state * 1103515245 + 12345) & 2147483647;
        long k = 1 + state % kmax;
        state = (state * 1103515245 + 12345) & 2147483647;
        long idx = state % n;
        prices[idx] = (state >> 16) % 1000;
        sink += max_profit(k, prices, n);
    }
    printf("%ld\n", sink);
    free(prices);
    return 0;
}

File: leetcode/101-200/188-best-time-to-buy-and-sell-stock-iv/bench/stock_iv.rs
fn max_i64(a: i64, b: i64) -> i64 {
    if a > b {
        a
    } else {
        b
    }
}

fn max_profit(k: i64, prices: &[i64]) -> i64 {
    let n = prices.len() as i64;
    if n == 0 || k == 0 {
        return 0;
    }
    if k >= n / 2 {
        let mut profit: i64 = 0;
        let mut i: i64 = 1;
        while i < n {
            if prices[i as usize] > prices[(i - 1) as usize] {
                profit += prices[i as usize] - prices[(i - 1) as usize];
            }
            i += 1;
        }
        return profit;
    }

    let neg: i64 = -1000000000;
    let mut buy: Vec<i64> = vec![neg; (k + 1) as usize];
    let mut sell: Vec<i64> = vec![0; (k + 1) as usize];
    let mut d: i64 = 0;
    while d < n {
        let price = prices[d as usize];
        let mut t: i64 = 1;
        while t <= k {
            buy[t as usize] = max_i64(buy[t as usize], sell[(t - 1) as usize] - price);
            sell[t as usize] = max_i64(sell[t as usize], buy[t as usize] + price);
            t += 1;
        }
        d += 1;
    }
    sell[k as usize]
}

fn main() {
    let n: i64 = 2000;
    let kmax: i64 = 50;
    let rounds: i64 = 5000;

    let mut prices: Vec<i64> = Vec::with_capacity(n as usize);
    let mut state: i64 = 12345;
    for _ in 0..n {
        state = (state * 1103515245 + 12345) & 2147483647;
        prices.push((state >> 16) % 1000);
    }

    let mut sink: i64 = 0;
    for _ in 0..rounds {
        state = (state * 1103515245 + 12345) & 2147483647;
        let k = 1 + state % kmax;
        state = (state * 1103515245 + 12345) & 2147483647;
        let idx = (state % n) as usize;
        prices[idx] = (state >> 16) % 1000;
        sink += max_profit(k, &prices);
    }
    println!("{}", sink);
}
